File: tests/FSharp.Compiler.ComponentTests/EmittedIL/FixedBindings/PinCSharpTypeWithGetPinnableReferenceReturningByref.fs
module FixedExpressions
open Microsoft.FSharp.NativeInterop
open System

let pinIt (thing: PinnableReference<int>) =
    use ptr = fixed thing
    NativePtr.get ptr 0
    
[<EntryPoint>]
let main _ =
    let x = PinnableReference(42)
    let y = pinIt x
    if y <> x.Value then failwith "y did not equal x value"
    0

File: tests/FSharp.Compiler.ComponentTests/EmittedIL/FixedBindings/PinCSharpTypeWithGetPinnableReferenceReturningByref.cs
public class PinnableReference<T>
{
    private T _value;

    public T Value
    {
        get => _value;
        set => _value = value;
    }

    public PinnableReference(T value)
    {
        this._value = value;
    }

    public ref T GetPinnableReference()
    {
        return ref _value;
    }
}